packages/react-native-reanimated/plugin/src/class.ts

```diff
@@ -0,0 +1,331 @@
+): boolean {
+  if (!classPath.node.id) {
+    // We don't support unnamed classes yet.
+    return false;
+  }
+
+  if (!hasWorkletClassMarker(classPath.node.body)) {
+    return false;
+  }
+
+  const parentPath = classPath.parentPath;
+  const className = classPath.node.id.name;
+
+  classPath.skip();
+
+  classPath = makeStatement(classPath, parentPath, className);
+
+  removeWorkletClassMarker(classPath.node.body);
+
+  processClass(classPath, state);
+
+  return true;
+}
+
+function processClass(
+  classPath: NodePath<ClassDeclaration>,
+  state: ReanimatedPluginPass
+) {
+  assert(classPath.node.id);
+  const className = classPath.node.id.name;
+  const code = generate(classPath.node).code;
+
+  const transformedCode = transformSync(code, {
+    plugins: [
+      '@babel/plugin-transform-class-properties',
+      '@babel/plugin-transform-classes',
+      '@babel/plugin-transform-unicode-regex',
+    ],
+    filename: state.file.opts.filename,
+    ast: true,
+    babelrc: false,
+    configFile: false,
+  });
+
+  assert(transformedCode);
+  assert(transformedCode.ast);
+
+  const ast = transformedCode.ast;
+
+  let factory: VariableDeclaration;
+
+  let hasHandledClass = false;
+
+  traverse(ast, {
+    [WorkletizableFunction]: {
+      // @ts-expect-error TS has some trouble inferring here.
+      enter: (functionPath: NodePath<WorkletizableFunction>) => {
+        if (functionPath.parentPath.isObjectProperty()) {
+          return;
+        }
+
+        const workletDirective = directive(directiveLiteral('worklet'));
+
+        if (functionPath.parentPath.isCallExpression() && !hasHandledClass) {
+          const factoryCopy = cloneNode(
+            functionPath.node,
+            true
+          ) as FunctionExpression;
+          factoryCopy.id = identifier(className + 'ClassFactory');
+          factoryCopy.body.directives.push(workletDirective);
+          factory = variableDeclaration('const', [
+            variableDeclarator(
+              identifier(className + 'ClassFactory'),
+              factoryCopy
+            ),
+          ]);
+          hasHandledClass = true;
+
+          return;
+        }
+
+        const bodyPath = functionPath.get('body');
+        if (!bodyPath.isBlockStatement()) {
+          bodyPath.replaceWith(
+            blockStatement([returnStatement(bodyPath.node as Expression)])
+          );
+        }
+
+        (functionPath.node.body as BlockStatement).directives.push(
+          workletDirective
+        );
+      },
+    },
+  });
+
+  const body = ast.program.body;
+  body.push(factory!);
+
+  body.push(
+    expressionStatement(
+      assignmentExpression(
+        '=',
+        memberExpression(
+          identifier(className),
+          identifier(className + 'ClassFactory')
+        ),
+        identifier(className + 'ClassFactory')
+      )
+    )
+  );
+
+  sortPolyfills(ast);
+
+  const transformedNewCode = transformSync(generate(ast).code, {
+    ast: true,
```

Comment:
Is the `ast` here necessary if we already have it?

Reply:
It wasn't needed anymore.